x-pack/legacy/plugins/advanced_ui_actions/public/np_ready/public/custom_time_range_action.tsx

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import React from 'react';
+import { TimeRange } from '../../../../../../../src/plugins/data/public';
+import { SEARCH_EMBEDDABLE_TYPE } from '../../../../../../../src/legacy/core_plugins/kibana/public/discover/embeddable/search_embeddable';
+import { VisualizeEmbeddable } from '../../../../../../../src/legacy/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable';
+import { VISUALIZE_EMBEDDABLE_TYPE } from '../../../../../../../src/legacy/core_plugins/kibana/public/visualize/embeddable/constants';
+
+import {
+  Action,
+  IEmbeddable,
+  ActionContext,
+  IncompatibleActionError,
+  Embeddable,
+  EmbeddableInput,
+} from '../../../../../../../src/legacy/core_plugins/embeddable_api/public/np_ready/public';
+
+import { CustomizeTimeRangeModal } from './customize_time_range_modal';
+import { OpenModal, CommonlyUsedRange } from './types';
+
+const CUSTOM_TIME_RANGE = 'CUSTOM_TIME_RANGE';
+
+export interface TimeRangeInput extends EmbeddableInput {
+  timeRange: TimeRange;
+}
+
+function hasTimeRange(
+  embeddable: IEmbeddable | Embeddable<TimeRangeInput>
+): embeddable is Embeddable<TimeRangeInput> {
+  return (embeddable as Embeddable<TimeRangeInput>).getInput().timeRange !== undefined;
+}
+
+function isVisualizeEmbeddable(
+  embeddable: IEmbeddable | VisualizeEmbeddable
+): embeddable is VisualizeEmbeddable {
+  return embeddable.type === VISUALIZE_EMBEDDABLE_TYPE;
+}
+
+export class CustomTimeRangeAction extends Action {
+  public readonly type = CUSTOM_TIME_RANGE;
+  private openModal: OpenModal;
+  private dateFormat?: string;
+  private commonlyUsedRanges: CommonlyUsedRange[];
+
+  constructor({
+    openModal,
+    dateFormat,
+    commonlyUsedRanges,
+  }: {
+    openModal: OpenModal;
+    dateFormat: string;
+    commonlyUsedRanges: CommonlyUsedRange[];
+  }) {
+    super(CUSTOM_TIME_RANGE);
+    this.order = 7;
+    this.openModal = openModal;
+    this.dateFormat = dateFormat;
+    this.commonlyUsedRanges = commonlyUsedRanges;
+  }
+
+  public getDisplayName() {
+    return i18n.translate('xpack.advancedUiActions.customizeTimeRangeMenuItem.displayName', {
+      defaultMessage: 'Customize time range',
+    });
+  }
+
+  public getIconType() {
+    return 'calendar';
+  }
+
+  public async isCompatible({ embeddable }: ActionContext) {
+    const isInputControl =
+      isVisualizeEmbeddable(embeddable) &&
+      embeddable.getOutput().visTypeName === 'input_control_vis';
+
+    const isMarkdown =
```

Comment:
I'm thinking we may want to re-think how this is implemented. Because we cannot hard-code here all incompatible vis-types, what if a plugin adds a new vis-type. Maybe instead of looking for incompatible vis-types an embeddable could have some property that resolves to `true` if time range affects its display.

Reply:
I had some thoughts awhile ago on how we could do this, but didn't want to over engineer until needed so put it on the back burner.  I was thinking of having something like an `ActionAdapters` registry that could be registered and used to connect actions to embeddables by third party plugin developers.  e.g person `A` creates Embeddable `1e`, person `B` creates action `2a`, person `C` could register an adapter that would allow action `2a` to work with embeddable `1e`.  Now that we are going to decouple actions from embeddables though, it'd need to be rethought.  But I could see something like that possibly working, where there is another registry that determines compatibility based on the trigger context + action, and the adapter has the opportunity to actually manipulate the trigger context to a shape that that action requires.  Would totally need to be thought through much more though.  

tl;dr; I agree with you but I would try to come up with something more generic than putting a fn directly on the embeddable interface for time range explicitly.